spec/did-method-spec-template.html

```diff
@@ -227,10 +173,111 @@ <h3>Read (Resolve)</h3>
+          rendering a DID Document as follows. See the notes following the rendering, and the next section that lists an earlier version
+          of the rendered document.
+        </p>
+
         <pre>
+          {
+            "@context": [
+              "https://www.w3.org/ns/did/v1"
+            ],
+            "id": "did:sov:HR6vs6GEZ8rHaVgjg2WodM",
+            "verificationMethod": [
+              {
+                "type": "Ed25519VerificationKey2018",
+                "id": "did:sov:HR6vs6GEZ8rHaVgjg2WodM#key-1",
+                "publicKeyBase58": "9wvq2i4xUa5umXoThe83CDgx1e5bsjZKJL4DEWvTP9qe"
+              }
+            ],
+            "service": [
+              {
+                "type": "endpoint",
+                "serviceEndpoint": "https://example.com/endpoint"
+              },
+              {
+                "id": "did:sov:HR6vs6GEZ8rHaVgjg2WodM#did-communication",
+                "type": "did-communication",
+                "priority" : 0,
+                "recipientKeys" : [ "did:sov:HR6vs6GEZ8rHaVgjg2WodM#key-1" ],
+                "routingKeys" : [ ],
+                "accept": [
+                  "didcomm/aip2;env=rfc19"
+                ],
+                "serviceEndpoint": "https://example.com/endpoint"
+              },
+              {
+                "id": "did:sov:HR6vs6GEZ8rHaVgjg2WodM#didcomm-1",
+                "type": "DIDComm",
+                "serviceEndpoint": "https://example.com/endpoint",
+                "routingKeys": [ ]
+              }
+            ],
+            "authentication": [
+              {
+                "type": "Ed25519VerificationKey2018",
+                "id": "did:sov:HR6vs6GEZ8rHaVgjg2WodM#key-1",
+                "publicKeyBase58": "9wvq2i4xUa5umXoThe83CDgx1e5bsjZKJL4DEWvTP9qe"
+              }
+            ],
+            "assertionMethod": [
+              {
+                "type": "Ed25519VerificationKey2018",
+                "id": "did:sov:HR6vs6GEZ8rHaVgjg2WodM#key-1",
+                "publicKeyBase58": "9wvq2i4xUa5umXoThe83CDgx1e5bsjZKJL4DEWvTP9qe"
+              }
+            ]
+          }
         </pre>
 
+        <h5>Notes</h5>
+
+        <p>The following are notes about the example DID Document above:
+          <ul>
+              <li>The DID Document text is boilerplate except for the "did:sov..." identifier, the public key ("publicKeyBase58) and the references to the service endpoint ("serviceEndpoint").</li>
+              <li>If there is no result from the GET_ATTRIB transaction, the "service" block is not included in the document.</li>
+              <li>The three entries in the service block cover the three "eras" of DIDComm, AIP 1.0 (type "endpoint"), AIP 2.0 (type "did-communication") and DIDComm V2 (type "DIDComm").</li>
+              <li>Because there is just one endpoint specified, the service blocks cannot make use of the DIDComm routing capabilities for messaging based on the DIDDoc.</li>
```

Comment:
>Because there is just one endpoint specified

I don't understand this comment. I see several different places where `serviceEndpoint` is given in the `service` array. All of them have the same value, but I don't understand why that would matter. Can the sentence be clarified?

Reply:
Daniel -- I'm trying not to limit what we're doing here.  Currently, the convention that Indy users are following is to have an ATTRIB with a `raw` value of `{"endpoint":{"endpoint":"http://example.com/didService"}}.  There are [313 such instances](https://indyscan.io/txs/SOVRIN_MAINNET/domain?page=1&pageSize=50&filterTxNames=[%22ATTRIB%22]&sortFromRecent=true&search=endpoint) (as I type this) on Sovrin MainNet today. 

If you look at the Universal Resolver, it's output includes that information, in the form of the simple "endpoint" service block.  What this PR does is just expand the single "endpoint" we have based on the convention being used, and provide proper DIDComm service blocks for AIP 2.0 and DIDComm V2.

Assuming we don't change the convention, we have no where to specify a mediator/routingKeys -- hence that is hard-wired to be an empty array.

What we really want is `did:indy` -- but that's not happening fast enough...

Make sense?

Reply:
Ah, yes. Thank you for the clarification. That helps a lot.

With this PR, would it be possible to start up a new convention (without writing any new code in the ledger--only in clients) where the `raw` value would also include `routingKeys` data, and this would then be returned by the UR? If so, then maybe it would be better to NOT insert an empty array for `routingKeys`, and just let the data stored by the ledger be returned. That way we could start defining endpoints without routing keys if we were motivated to do so.

Reply:
I'd really like to the see the focus put on did:indy for that, so I personally would prefer not.  I suppose it would take off the pressure to get did:indy implemented.  I'd love to know how much people would use this if we did it now.  Easy enough though -- like this I think:

``` json
{
  "endpoint": {
    "endpoint": "http://example.com/didService",
    "routingKeys": [ "did:key:12345", "did:key:67890" ]
  }
}

```

BTW -- my reply to you above was terrible.  Just noticed my opening line was the opposite of intended.  I said "I'm trying not to limit", but meant to say "I'm trying ~not~ to limit".  Doh :rofl

Reply:
I see the logic of limiting coding effort -- but what I'm proposing is not extra coding. It's actually less coding. Instead of hard-coding some logic into UR that always adds an empty `routingKeys`, I'm suggesting that we just ignore the issue entirely. If someone writes an ATTRIB that includes `routingKeys` inside `raw`, then UR returns it because the ledger returns it. If they don't have an ATTRIB with `routingKeys`, UR doesn't emit one, either. All the current ATTRIBs on the record would lack `routingKeys` and thus be invalid DIDComm endpoints -- and so would new ones written to the ledger using the status quo convention. But by doing nothing, we'd make it possible to adopt a new convention if we were forced to.

My suggestion is motivated by a lack of confidence in the timeline for did:indy. If we had did:indy today, I wouldn't suggest it at all, because we'd have a better alternative. Perhaps I'm being pessimistic to want to leave us a pressure release valve. Or perhaps there are reasons not to do this that I haven't considered. I wasn't advocating super strongly -- just thinking out loud. I'm content to have the PR merged as-is, if others feel that's the best tradeoff.

Reply:
It's not as simple as you say.  First, the resolver has to know about each `raw` type (e.g. "endpoint", "routingKeys") and specifically make a call to the ledger to ask for each value.  There is no way to know what ATTRIBs a client has written and we don't want the resolver to have to ask for a whole bunch of ATTRIBs on the off chance a DID controller used a specific set of ATTRIBs.  As such, I would not want to add any other ATTRIBs as part of this process.  "did:indy" is needed for that.

As noted, we *could* add an extra format for "endpoint" to add `routingKeys`, but the resolvers would have to add support for that format (update all existing deployments) and clients would have to know about it to use it.  Since these endpoints are likely used only during connection establishment, and will be replaced by peer DIDs, the lack of mediators could be worked around.

I agree that "did:indy" won't be around for awhile. However, I think this will be sufficient for the institutional issuers we have today -- the main use case we have for DIDs on Indy.

Reply:
Ah, got it. I had forgotten that each `raw` value was a separate call. Then I agree with your conclusion.